Use geometry to define shape of distributions

nabriis opened the pull request:
Closes #28, Closes #212

nabriis commented:
@amal-ghamdi, @jakobsj  I believe this is ready for review now. It adds the option for distributions to use geometry (or dim) to define shape. Most changes are in adding a significant number of tests in the code.

Finally there was a decision to make related to GMRF, LMRF, CMRF. Should we use consistent interface with a location parameter. In the end I decided to do it (in the paper we compare all 3 in a single cell using the location parameter in section 3).

This means GMRF, LMRF and CMRF all can be defined with a location and 2nd parameter e.g.

```python
x_zero = np.zeros(n)

x = GMRF(x_zero, 1) # or
x = LMRF(x_zero, 1) # or
x = CMRF(x_zero, 1)
```

In addition the following would also be possible

```python

x = GMRF(0, 1, dim=n) # or
x = LMRF(0, 1, dim=n) # or
x = CMRF(0, 1, dim=n)
```

GMRF needs to be updated to fully support this, but I decided this was perhaps best for another PR.

amal-ghamdi commented on cuqi/distribution/_distribution.py:
```diff
@@ -84,42 +84,69 @@ def __init__(self, name=None, geometry=None, is_symmetric=None):
+        inferred_dim = self._infer_dim_of_mutable_variables()
+
+        # If inconsistent geometry dimensions, raise an exception
+        if inferred_dim and self._geometry.par_dim and inferred_dim > 1 and self._geometry.par_dim != inferred_dim:
```
Just wanted to ask about the `inferred_dim` > 1 condition, is there a reason to exclude raising this exception when inferred_dim ==1?.

nabriis replied:
Updated to improve readability

amal-ghamdi commented on cuqi/distribution/_lmrf.py:
```diff
@@ -39,35 +40,52 @@ class LMRF(Distribution):
         self._bc_type = bc_type
+
+        # Ensure geometry has shape
+        if not self.geometry.fun_shape or self.geometry.par_dim == 1:
```
Just wondering if this should be `and` condition instead of `or` (same above in line 60 in _cmrf.py). The if statement does not seem to match `must be initialized with geometry or dim greater than 1.`. Could be something I am missing.

Another related point, FEniCS geometries has fun_shape None, so this error will be raised always.

nabriis replied:
Added more details in error statement

amal-ghamdi commented on cuqi/sampler/_laplace_approximation.py:
```diff
@@ -130,9 +130,15 @@ def Lk_fun(x_k):
+        if len(self.target.prior.location) == 1:
+            self._priorloc = np.repeat(self.target.prior.location, self.dim)
+        else:
+            self._priorloc = self.target.prior.location
```
Just a suggestion to make `self._priorloc` not class variable but only defined locally in the method e.g.  `_priorloc` instead of  `self._priorloc`. Because it is not used in other methods in this calss, and just in case we provide flexibility to update the prior then this can prevent inconsistency.

nabriis replied:
Most other parameters were stored like that. I agree perhaps all should be updated to be defined in the sample scope, but for now I suggest we keep it like this.

amal-ghamdi commented on tests/test_distributions_shape.py:
```diff
@@ -0,0 +1,103 @@
+@pytest.mark.parametrize("dist", dists, ids=lambda d: d.__name__)
+def test_multivariate_scalar_vars_logd(dist):
+    if dist in skip_logd:
+        pytest.skip(f"{dist.__name__} does not handle scalar setup for logd yet.")
```
can create an issue to capture these cases. And similarly in the two tests below

nabriis replied:
Yes!

nabriis replied:
# Define cases to skip (these are TODO)
skip_logd = [
    cuqi.distribution.GMRF, # Mean must allow scalar values
    cuqi.distribution.Gamma # Missing force_ndarray
]
skip_sample = [
    cuqi.distribution.GMRF,  # Mean must allow scalar values
    cuqi.distribution.Gamma, # Missing force_ndarray
    cuqi.distribution.Lognormal,
]
skip_gradient = [
    cuqi.distribution.GMRF,  # Mean must allow scalar values
    cuqi.distribution.Gamma, # Missing force_ndarray
    cuqi.distribution.Lognormal,
    cuqi.distribution.InverseGamma, 
]

I would also suggest adding an issue to have some default variable parameters to initialize the distributions with.

amal-ghamdi commented on tests/test_distributions_shape.py:
```diff
@@ -0,0 +1,103 @@
+import cuqi
```
This is such a cool testing framework :)

amal-ghamdi reviewed (commented):
Thank you @nabriis for adding these great features to increase flexibility, consistency, and ease of use of defining distributions!

I added just few comments.

jakobsj commented on cuqi/distribution/_cmrf.py:
```diff
@@ -46,31 +48,50 @@ class CMRF(Distribution):
     """
    
-    def __init__(self, location, scale, bc_type="zero", physical_dim=1, **kwargs):
+    def __init__(self, location, scale, bc_type="zero", physical_dim=None, **kwargs):
```
Not sure, but I had imagined physical_dim would no longer be exposed?

jakobsj replied:
I suppose is still needed if no geometry is passed?

jakobsj commented on README.md:
```diff
@@ -34,11 +34,11 @@ from cuqi.problem import BayesianProblem
-                 bc_type='neumann',
-                 physical_dim=2)
+# Step 2: Prior, x ~ LMRF(0,0.01)
+x = LMRF(location=0,
```
In the updated, a scalar zero is given for location, in combination with the geometry, will zeros be filled for the location values of all the elements then? What is a numpy 1D array is given for the location and the size is not compatible with the geometry?  I understood from you that all distributions would now also allow passing a `dim` parameter, does this include XMRF, and if so how are the possibly conflicting ways of passing input handled?

jakobsj commented on demos/demo15_test_LMRF.py:
```diff
@@ -7,12 +7,12 @@
 dom = 1
 BCs = 'neumann'
-x = cuqi.distribution.LMRF(mean, prec, BCs)
+x = cuqi.distribution.LMRF(location, prec, BCs, dim=N)
```
Here a 1D version is obtained. I want to understand which combinations of location, dim, physical_dim and geometry are allowed, incl in which different ways one can set up a 2D XMRF. 

Can you do a scalar location/prec, a scalar dim and physical_dim = 1 or 2?
Can you do a scalar location/prec, no dim, and a physical_dim 1 or 2?
Can you do scalar location/prec, no dim, no physical_dim, and a 1D or 2D geometry?
Can you do vec location/prec, no dim, no physical_dim, and a 1D or 2D geometry?
etc

And if all conflicting sets of inputs are handled? 

Also what happens if geometry that doesn't make sense with XMRF is passed, like StepExpansion?

jakobsj reviewed (commented):
Looks great @nabriis, I added some quick comments/questions inline.

nabriis commented:
We discussed to ensure that geometry is enforced at distribution creation. Either inferred via parameters or from explicit geometry= keyword.

We agreed to remove dim keyword and physical_dim keyword

nabriis commented:
@jakobsj  could you review?

Removed physical dim and dim keywords from distribution. Updated so default geometry has a 2D version. Added consistency check for Gaussian so that it cannot be defined without a geometry.

The remainder of updates can perhaps be reviewed by @amal-ghamdi once added.

jakobsj reviewed (approved):
Quick approval, looks very good, but didn't check details or try out.  Using geometry as input looks good to me, I think it is the right decisions. Merge pending more in-depth review and approval from @amal-ghamdi

nabriis commented:
Hi @amal-ghamdi, this is ready for review now.

amal-ghamdi commented on tests/test_MRFs.py:
```diff
@@ -1,55 +1,69 @@
 def test_LMRF_default_geometry_is_replaced_when_physical_dim_is_2():
-    lmrf = LMRF(0, scale=0.2, physical_dim=2, dim=16)
+    lmrf = LMRF(0, scale=0.2, geometry=(4,4))
     assert isinstance(lmrf.geometry, Image2D)
```
can assert Default2D

nabriis replied:
Good point!

amal-ghamdi commented on tests/test_geometry.py:
```diff
@@ -388,3 +388,5 @@ def test_KLExpansion_projection():
     rel_err = np.linalg.norm(signal-signal_proj)/np.linalg.norm(signal)
     assert np.isclose(rel_err, 0.0, atol=1e-5)
+
+# TODO. Add tests for _DefaultGeometry2D and add tests for default initialiations
```
This can be added as an issue: "Add tests for _DefaultGeometry2D and add tests for default initializations"

nabriis replied:
Added a basic test

amal-ghamdi commented on tests/test_posterior.py:
```diff
@@ -0,0 +1,36 @@
+    x = cuqi.distribution.Gaussian(0, 1, geometry=cuqi.geometry.Image2D((10, 10)))
+    y = cuqi.distribution.Gaussian(A(x), 1)
+    posterior = cuqi.distribution.Posterior(y.to_likelihood(np.zeros(10**2)), x)
+    assert posterior.geometry == x.geometry
```
I looked a bit into this and it seems, `posterior.geometry` is `_DefaultGeometry2D(100,)` while is x.geometry is `Image2D(100,)`. Is this the expected behavior, or is the expected behavior that `posterior.geometry` is `Image2D(100,)`. 

This could be related to how we sat the equivalence logic in geometries (_DefaultGeometry2D is image2D as well).

nabriis replied:
Great catch. Fixed the bug

amal-ghamdi commented on tests/test_posterior.py:
```diff
@@ -0,0 +1,36 @@
+    x = cuqi.distribution.Gaussian(0, 1, geometry=(10,10))
+    y = cuqi.likelihood.UserDefinedLikelihood(logpdf_func=lambda x: x, geometry=cuqi.geometry.Image2D((10, 10)))
+    posterior = cuqi.distribution.Posterior(y, x)
+    assert posterior.geometry == y.geometry
```
also here posterior geometry is `_DefaultGeometry2D(100,)`

nabriis replied:
Same here. Fixed it

amal-ghamdi reviewed (commented):
Thank you @nabriis for this great feature upgrade!

I left some comments here and there for your consideration.

nabriis commented on cuqi/distribution/_cmrf.py:
```diff
@@ -1,22 +1,23 @@
+    For 2D the differences are defined in both horizontal and vertical directions.
 
     It is possible to define boundary conditions using the `bc_type` parameter.
 
```
Done

nabriis commented on cuqi/distribution/_cmrf.py:
```diff
@@ -46,31 +47,43 @@ class CMRF(Distribution):
+
+        # Ensure geometry has shape
+        if not self.geometry.fun_shape or self.geometry.par_dim == 1:
+            raise ValueError(f"Distribution {self.__class__.__name__} must be initialized with supported geometry (with fun_shape) or dim greater than 1.")
```
Done

nabriis commented on cuqi/distribution/_distribution.py:
```diff
@@ -377,6 +405,12 @@ def _parse_args_add_to_kwargs(self, cond_vars, *args, **kwargs):
+    
+    def check_geometry_consistency(self):
+        """ Checks that the geometry of the distribution is consistent with inferred dimension """
+
```
This is a good point. Also related to the comment on geometry consistency you gave [here](https://github.com/CUQI-DTU/CUQIpy/pull/246#discussion_r1250792902). If we want to enforce a geometry to be set, we need to check it after all subclasses have set their properties. E.g. if a Gaussian is made with Gaussian(0,1) we can assume geometry can be _DefaultGeometry(1,) which is inferred when calling .geometry. This needs to be checked at the end of subclass init.

nabriis commented on cuqi/distribution/_gaussian.py:
```diff
@@ -118,6 +118,8 @@ def __init__(self, mean=None, cov=None, prec=None, sqrtcov=None, sqrtprec=None,
             self._mutable_vars = ['mean', 'sqrtprec']
             self.sqrtprec = sqrtprec
 
+        self.check_geometry_consistency()
```
See https://github.com/CUQI-DTU/CUQIpy/pull/246#discussion_r1252348420

nabriis commented on cuqi/distribution/_gmrf.py:
```diff
@@ -2,7 +2,7 @@
 from scipy.sparse import diags, eye
```
Done see #255

nabriis commented on cuqi/distribution/_lmrf.py:
```diff
@@ -1,20 +1,21 @@
-    For 2D `(physical_dim=2)` the differences are defined in both horizontal and vertical directions.
+    For 2D the differences are defined in both horizontal and vertical directions.
 
     It is possible to define boundary conditions using the `bc_type` parameter.
```
Done

nabriis commented on cuqi/distribution/_lmrf.py:
```diff
@@ -39,35 +40,45 @@ class LMRF(Distribution):
+
+        # Ensure geometry has shape
+        if not self.geometry.fun_shape or self.geometry.par_dim == 1:
+            raise ValueError(f"Distribution {self.__class__.__name__} must be initialized with supported geometry (with fun_shape) or dim greater than 1.")
```
Done

nabriis commented on cuqi/model/_model.py:
```diff
@@ -124,8 +126,10 @@ def __init__(self, forward, range_geometry, domain_geometry, gradient=None, jaco
+            self.domain_geometry = _DefaultGeometry1D(grid=domain_geometry)
         elif isinstance(domain_geometry, Geometry):
             self.domain_geometry = domain_geometry
         elif domain_geometry is None:
```
Done

nabriis commented:
Hi @amal-ghamdi. Took your comments into account.

amal-ghamdi reviewed (approved):
Thank you @nabriis for the updates. LGTM!

I approved and added two comments about minor typo.
